sources/csv/data.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csv
+
+import (
+	csvReader "encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"strconv"
+	"time"
+
+	"cloud.google.com/go/civil"
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+)
+
+type Column struct {
+	Column_name string `json:"column_name"`
+	Type_name   string `json:"type_name"`
+}
+
+// Harbourbridge accepts a manifest file in the form of a json which unmarshalls into the Table struct.
+type Table struct {
+	Table_name    string   `json:"table_name"`
+	File_patterns []string `json:"file_patterns"`
+	Columns       []Column `json:"columns"`
+}
+
+// LoadManifest reads the manifest file and unmarshalls it into a list of Table struct.
+// It also performs certain checks on the manifest.
+func LoadManifest(manifestFile string) ([]Table, error) {
+	manifest, err := ioutil.ReadFile(manifestFile)
+	if err != nil {
+		return nil, fmt.Errorf("can't read manifest file due to: %v", err)
+	}
+	tables := []Table{}
+	err = json.Unmarshal(manifest, &tables)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshall json due to: %v", err)
+	}
+	err = verifyManifest(tables)
+	if err != nil {
+		return nil, fmt.Errorf("manifest is incomplete: %v", err)
+	}
+	return tables, nil
+}
+
+// verifyManifest performs certain prechecks on the structure of the manifest.
+// Checks on valid file paths and empty CSVs are handled as conv.Unexpected errors later during processing.
+func verifyManifest(tables []Table) error {
+	if len(tables) == 0 {
+		return fmt.Errorf("no tables found")
+	}
+	for i, table := range tables {
+		name := table.Table_name
+		if name == "" {
+			return fmt.Errorf("table number %d (0-indexed) does not have a name", i)
+		}
+		if len(table.File_patterns) == 0 {
+			return fmt.Errorf("no file path provided for table %s", name)
+		}
+		cols := table.Columns
+		if len(cols) == 0 {
+			return fmt.Errorf("`columns` field for table %s is empty", name)
+		}
+		for j, col := range cols {
+			if col.Column_name == "" || col.Type_name == "" {
```

Comment:
We should check for valid column types supported here before we start the migration.